File: exercise/ts/day20/src/service.ts
import {v4 as uuidv4} from 'uuid';
import {Either, left, right} from 'fp-ts/lib/Either';
import {Reindeer, ReindeerColor, ReindeerErrorCode} from './types';

export class ReindeerService {
    private reindeer: Reindeer[] = [
        new Reindeer("40f9d24d-d3e0-4596-adc5-b4936ff84b19", 'Petar', ReindeerColor.Black)
    ];

    public get(id: string): Either<ReindeerErrorCode, Reindeer> {
        const reindeer = this.reindeer.find(r => r.id === id);
        return reindeer ? right(reindeer) : left(ReindeerErrorCode.NotFound);
    }

    public create(reindeerToCreate: ReindeerToCreate): Either<ReindeerErrorCode, Reindeer> {
        const existingReindeer = this.reindeer.find(r => r.name === reindeerToCreate.name);
        if (existingReindeer) return left(ReindeerErrorCode.AlreadyExist);

        const newReindeer = new Reindeer(uuidv4(), reindeerToCreate.name, reindeerToCreate.color);
        this.reindeer.push(newReindeer);

        return right(newReindeer);
    }
}

export class ReindeerToCreate {
    constructor(
        public name: string,
        public color: ReindeerColor
    ) {
    }
}

File: exercise/python/day20/reindeer/service.py
from typing import List, Optional
from uuid import uuid4


class ReindeerColor:
    WHITE = 0
    BLACK = 1
    PURPLE = 2


class Reindeer:
    def __init__(self, id: str, name: str, color: int):
        self.id = id
        self.name = name
        self.color = color


class ReindeerErrorCode:
    NOT_FOUND = "NotFound"
    ALREADY_EXIST = "AlreadyExist"


class ReindeerService:
    def __init__(self):
        self.reindeer: List[Reindeer] = [
            Reindeer("40f9d24d-d3e0-4596-adc5-b4936ff84b19", "Petar", ReindeerColor.BLACK)
        ]

    def get(self, id: str) -> Optional[Reindeer]:
        for reindeer in self.reindeer:
            if reindeer.id == str(id):
                return reindeer
        return None

    def create(self, name: str, color: int) -> Optional[str]:
        for reindeer in self.reindeer:
            if reindeer.name == name:
                return ReindeerErrorCode.ALREADY_EXIST
        new_reindeer = Reindeer(str(uuid4()), name, color)
        self.reindeer.append(new_reindeer)
        return new_reindeer.id
